doc/TACACSPLUS_PASSKEY_ENCRYPTION.md

```diff
@@ -0,0 +1,105 @@
+3. When TACACS+ Authentication is activated on the device, a new PAM configuration file (common-auth-sonic) is generated and substituted in the login and SSH daemons. Importantly, the pre-existing configuration file remains unchanged.
+The revised data handling procedure among the modules is outlined as follows:
+1. When a user configures the TACACS passkey using the SONIC CLI, it will now be securely stored in encrypted format instead of plaintext.
+2. Subsequently, the HostCfg Enforcer module retrieves this encrypted key. However, before writing it into the PAM configuration file(s), the hostCfgd module decrypts it.
+```
+       +-------+  +---------+
+       |  SSH  |  | Console |
+       +---+---+  +----+----+
+           |           |   
++----------v-----------v---------+                                      +---------------------+
+| AUTHENTICATION                 |                                      |                     |
+|   +-------------------------+  |         Decrypted passkey            |  +------------+     |
+|   | PAM Configuration Files    <------------+       +-----------------+  | AAA Config |     |
+|   +-------------------------+  |            |       |                 |  +------------+     |
+|                                |            |       |                 |                     |
+|         +-------------+        |         +-------------+              |  HostCfg Enforcer   |
+|         |     PAM     |        |         |             |   key-store  +----------^----------+
+|         |  Libraries  |        |    +---->  Master key |     __                  |
+|         +-------------+        |    |    |   Manager   |----|r |                 | Encrypted passkey
++---------------+----------------+    |    |             |    |o |                 |
+                |                     |    +-------------+    |o |                 |
+           +----v----+                |           |           |t |         +-------+--------+
+           |         |                |           |            --          |                |
+           |   CLI   +----------------+           +------------------------>    ConfigDB    |
+           |         |                                  Encrypted passkey  |                |
+           +---------+                                                     +----------------+
+```
+This decryption step is crucial because the login or SSH daemon references the PAM config file to verify the TACACS secret / passkey. If it remains encrypted, the SSH daemon will be unable to recognize the passkey, leading to login failures. The depicted block diagram clearly showcases the enhanced capabilities of the existing submodules.
+
+
+### Approach
+1. A runtime flag (via config_db) for Enable / disable feature: "key_encrypt"
```

Comment:
@liuh-80 I have already updated the yang model in the [PR](https://github.com/sonic-net/sonic-buildimage/pull/17201). I will also add the new schema in this HLD.

Reply:
The passkey encryption in CONFIG_DB will be done during the configuration part (in aaa.py). 
No Liuh, we don't need to have 'key_encrypt' flag in the existing schema. It is optional. If this flag is absent, encryption will not happen and the passkey will be pushed in plain text format only inside CONFIG_DB.

After rollout, if the user wants to enable this feature, he / she needs to simply enable "key_encrypt" flag and re-configure the passkey so that it will get encrypted and further replaced the same in CONFIG_DB.

Reply:
When user enable the "key_encrypt" flag, he / she needs to reconfigure the passkey. This is the requirement. Two steps are required to be followed if anyone wants to use this feature. But this will certainly not hamper the back-compatibility. It is backward compatible.

Reply:
It is like, if we need to configure SVI, we first need to add a vlan interface and later mark it as L3 interface by assigning an IP. Two steps. :-)

Reply:
BTW, if we are using CLI to encrypt the passkey, it will take care of updating the same into CONFIG_DB.

Reply:
You want me to add CLI to enable 'key_encrypt' flag for TACACS?

Reply:
@liuh-80 I have now provided an option to set the key_encrypt flag in TACPLUS table via CLI. Please check the [PR-Part I](https://github.com/sonic-net/sonic-utilities/pull/3027).  Now user need to provide -e option for enabling encryption feature. If not used, it will internally mark key_encrypt flag as false. Thus, post rollout of this feature, that flag will automatically be added in during the configuration itself. 

```
sonic/# config tacacs passkey -h
Usage: config tacacs passkey [OPTIONS] <secret_string>

  Specify TACACS+ server global passkey <STRING>

Options:
  -e, --encrypt   Enable passkey encryption feature
  -h, -?, --help  Show this message and exit.
```

And as mentioned in the previous message, hostcfgd will take care of handling the backward compatibility by doing the decryption only when key_encrypt flag is present and it is set to true. Please refer [PR - Part II](https://github.com/sonic-net/sonic-host-services/pull/81)